autometa/common/kmers.py

```diff
@@ -245,32 +338,39 @@ def normalize(df):
         index='contig', cols=[kmer, kmer, ...]
         Columns have been transformed by CLR normalization.
     """
-    # OPTIMIZE: May be able to implement this transformation with dask?
-    return df.dropna(axis='columns', how='all')\
-        .transform(lambda x: (x+1) / x.sum(), axis='columns')\
-        .transform(lambda x: np.log(x / gmean(x)), axis='columns')
-
-
-def embed(kmers=None, embedded=None, n_components=2, do_pca=True, pca_dimensions=50, method='TSNE', perplexity=30, **kwargs):
+    # steps in 1: data cleaning
+    df.dropna(axis='columns', how='all', inplace=True)
+    df.dropna(axis='index', how='all', inplace=True)
+    df.fillna(0, inplace=True)
+    # steps in 2 and 3: normalization and CLR transformation
+    step_2a = lambda x: (x+1) / x.sum()
+    step_2b = lambda x: np.log(x / gmean(x))
+    return df.transform(step_2a, axis='columns').transform(step_2b, axis='columns')
+
+def embed(kmers=None, embedded=None, n_components=2, do_pca=True, pca_dimensions=50,
+    method='UMAP', perplexity=30.0, **kwargs):
     """Embed k-mers using provided `method`.
 
     Parameters
     ----------
     kmers : str or pd.DataFrame
-        Description of parameter `kmers` (the default is None).
+        </path/to/input/kmers.normalized.tsv>
     embedded : str
-        Description of parameter `embedded` (the default is None).
+        </path/to/output/kmers.embedded.tsv> [optional] If provided will write to `embedded`.
     n_components : int
-        `n_components` to embed k-mer frequencies (the default is 3).
+        `n_components` to embed k-mer frequencies (the default is 2).
     do_pca : bool
         Perform PCA decomposition prior to embedding (the default is True).
     pca_dimensions : int
         Reduce k-mer frequencies dimensions to `pca_dimensions` (the default is 50).
         If None, will estimate based on
     method : str
-        Description of parameter `method` (the default is 'TSNE').
+        embedding method to use (the default is 'UMAP').
     perplexity : float
-        Description of parameter `perplexity` (the default is 30).
+        hyperparameter used to tune TSNE (the default is 30.0).
+        See below for details:
+            https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html#sklearn.manifold.TSNE
```

Comment:
You could also reference our NAR paper, because there are a couple of paragraphs on the perplexity setting.

Reply:
Going into detail here may be more appropriate for the readthedocs documentation. Although, when we have that up, we can reference that in the code. I'll put a COMBAK statement there for now